test/cpp_api_parity/module_impl_check.py

```diff
@@ -0,0 +1,308 @@
+    # We can do this because we are only interested in transferring
+    # the Python module's parameters and buffers to the C++ module.
+    module.forward = types.MethodType(lambda self, input: input, module)
+    script_module = torch.jit.trace(module, torch.tensor(0))
+
+    # Backward pass
+    python_output.sum().backward()
+
+    # Put all gradients into a dict, to be compared later
+    python_grad_dict = {}
+    for name, param in module.named_parameters():
+        grad = param.grad
+        if grad.is_sparse:
+            python_grad_dict[name + "_grad_indices"] = grad.coalesce().indices()
+            python_grad_dict[name + "_grad_values"] = grad.coalesce().values()
+        else:
+            python_grad_dict[name + "_grad"] = grad
+
+    return script_module, python_output, python_grad_dict
+
+def test_forward_backward(unit_test_class, test_params):
+    module_variant_name = test_params.module_variant_name
+    cpp_tmp_folder = test_params.cpp_tmp_folder
+
+    # Run forward and backward on Python module
+    script_module, python_output, python_grad_dict = run_python_forward_backward(unit_test_class, test_params)
+
+    # Save Python module and arguments to be used from C++ function
+    module_file_path = compute_temp_file_path(cpp_tmp_folder, module_variant_name, 'module')
+    arg_dict_file_path = compute_temp_file_path(cpp_tmp_folder, module_variant_name, 'arg_dict')
+    script_module.save(module_file_path)
+    serialize_arg_dict_as_script_module(test_params.arg_dict).save(arg_dict_file_path)
+
+    cpp_test_name = '{}_{}'.format(test_params.module_variant_name, 'test_forward_backward')
+    cpp_test_fn = getattr(unit_test_class.module_impl_check_cpp_module, cpp_test_name)
+
+    def run_cpp_test_fn_and_check_output():
+        forward_output_file_path = compute_temp_file_path(cpp_tmp_folder, module_variant_name, 'forward_output')
+        backward_grad_dict_file_path = compute_temp_file_path(cpp_tmp_folder, module_variant_name, 'backward_grad_dict')
+
+        cpp_test_fn(arg_dict_file_path, module_file_path, forward_output_file_path, backward_grad_dict_file_path)
+        cpp_output = torch.load(forward_output_file_path)
+        cpp_grad_dict = torch.load(backward_grad_dict_file_path)
+
+        # Check that forward outputs are equal
+        unit_test_class.assertEqual(python_output, cpp_output,
+                                    message=generate_error_msg("forward output", cpp_output, python_output))
+
+        # Check that module parameter gradients are equal after backward pass
+        unit_test_class.assertEqual(
+            len(python_grad_dict), len(cpp_grad_dict),
+            message=generate_error_msg("# of parameters", len(cpp_grad_dict), len(python_grad_dict)))
+        for key in python_grad_dict:
+            unit_test_class.assertTrue(
+                key in cpp_grad_dict,
+                msg=generate_error_msg("\"Does module have a parameter named `{}`?\"".format(key[:-5]), False, True))
```

Comment:
nit: Should comment on where the -5 is coming from. This would probably be nice to do via asserting that all keys are in the form `grad_*`